src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,247 @@
+               sqlStorage.commitTransaction(con);
+               return keys;
+            });
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            List<JWTSigningKeyInfo> keys;
+            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+            if (keysFromStorage == null) {
+                keys = new ArrayList<>();
+            } else {
+                keys = keysFromStorage;
+            }
+
+            return keys;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    /**
+     * Used to retrieve a key for JWT validation, for a given signing algorithm. If there are no keys in storage that match the given algorithm a new one is generated.
+     * @param algorithm The signing algorithm
+     * @return {@link JWTSigningKeyInfo} key
+     * @throws UnsupportedJWTSigningAlgorithmException If there is an error in the provided algorithm when getting/generating keys
+     * @throws StorageQueryException If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     * @throws DuplicateKeyIdException If there is an error setting generated keys in storage
+     */
+    public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms algorithm)
+            throws UnsupportedJWTSigningAlgorithmException, StorageQueryException, StorageTransactionLogicException, DuplicateKeyIdException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            try {
+                return sqlStorage.startTransaction(con -> {
+                    JWTSigningKeyInfo keyInfo = null;
+
+                    List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+                    // If there are no keys in storage, generate a new one
+                    if (keysFromStorage.isEmpty()) {
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                        } catch (NoSuchAlgorithmException | DuplicateKeyIdException | UnsupportedJWTSigningAlgorithmException e) {
+                            throw new StorageTransactionLogicException(e);
+                        }
+                    } else {
+                        // Loop through the keys and find the first one for the algorithm
+                        for (int i = 0; i < keysFromStorage.size(); i++) {
+                            JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                            if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
```

Comment:
Should you just implement a `.equals` function in the algorithm class so that you can do `currentKey.algorithm.equals(algorithm)`

Reply:
Makes sense, but `.equals` implies matching with the same type I feel so using `.equalsString` instead